### ISSUE TITLE
Configs are Mutated by File-Level Overrides

### ISSUE THREAD

**Issue by brandonschabell** [sqlfluff/sqlfluff#6343]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

When using a `FluffConfig` object as a config when linting a SQL string, any file-level overrides will mutate the config object, causing subsequent lintings to use the same file-level override. 

### Expected Behaviour

```python
def test__api__immutable_config():
    config = FluffConfig.from_path("test/fixtures/api/config_path_test/extra_configs/.sqlfluff")
    assert config.get("dialect") == "ansi"
    sqlfluff.lint(
        "-- sqlfluff:dialect: postgres\n"
        "SELECT * FROM table1\n",
        config=config
    )
    assert config.get("dialect") == "ansi"  # dialect should be unchanged
```

### Observed Behaviour

The test above fails:
`AssertionError: assert 'postgres' == 'ansi'`
The file-level override mutates the config object.

### How to reproduce

This test reproduces the issue:
```python
def test__api__immutable_config():
    config = FluffConfig.from_path("test/fixtures/api/config_path_test/extra_configs/.sqlfluff")
    assert config.get("dialect") == "ansi"
    sqlfluff.lint(
        "-- sqlfluff:dialect: postgres\n"
        "SELECT * FROM table1\n",
        config=config
    )
    assert config.get("dialect") == "ansi"
```

### Dialect

Ansi & Postgres, but this has been observed with every dialect.

### Version

3.2.3

### Configuration

I am using the test config available at `test/fixtures/api/config_path_test/extra_configs/.sqlfluff`.

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
